File: libUtil/src/main/java/org/waynezhou/libUtil/event/EventHandler.java
package org.waynezhou.libUtil.event;

import androidx.annotation.NonNull;

public class EventHandler<TEventArgs> {
    @NonNull
    private final EventHolder<TEventArgs> holder = new EventHolder<>();
    
    @NonNull
    public EventHolder<TEventArgs>.ListenerToken on(@NonNull EventListener<TEventArgs> listener) {
        return holder.addOnListener(listener);
    }
    
    @NonNull
    public EventHolder<TEventArgs>.ListenerToken once(@NonNull EventListener<TEventArgs> listener) {
        return holder.addOnceListener(listener);
    }
    
    public void removeListener(@NonNull EventListener<TEventArgs> listener) {
        holder.removeListener(listener);
    }
    
    public void removeListener(@NonNull EventHolder<TEventArgs>.ListenerToken token) {
        holder.removeListener(token);
    }
    
    public void invoke(TEventArgs e) {
        holder.invoke(e, () -> false);
    }
    
    public void invoke(
      TEventArgs e,
      @NonNull Condition breakAction
    ) {
        holder.invoke(e, breakAction);
    }
}

File: libUtilKt/src/main/java/org/waynezhou/libutilkt/event/EventHandler.kt
package org.waynezhou.libutilkt.event

class EventHandler<TEventArgs> {
    private val holder = EventHolder<TEventArgs>()

    fun on(listener: EventListener<TEventArgs>): EventHolder<TEventArgs>.ListenerToken {
        return holder.addOnListener(listener)
    }

    fun once(listener: EventListener<TEventArgs>): EventHolder<TEventArgs>.ListenerToken {
        return holder.addOnceListener(listener)
    }

    fun removeListener(listener: EventListener<TEventArgs>) {
        holder.removeListener(listener)
    }

    fun removeListener(token: EventHolder<TEventArgs>.ListenerToken) {
        holder.removeListener(token)
    }

    fun invoke(e: TEventArgs, breakAction: ()->Boolean = {false}) {
        holder.invoke(e, breakAction)
    }
}
